crates/mun_hir/src/method_resolution.rs

```diff
@@ -0,0 +1,306 @@
+    pub(crate) fn add_module_diagnostics(
+        &self,
+        db: &dyn HirDatabase,
+        module_id: LocalModuleId,
+        sink: &mut DiagnosticSink<'_>,
+    ) {
+        self.diagnostics
+            .iter()
+            .filter(|it| it.module_id(db.upcast()).local_id == module_id)
+            .for_each(|it| it.add_to(db, sink));
+    }
+
+    /// Adds all the `InherentImplsDiagnostics`s of the result to the `DiagnosticSink`.
+    pub(crate) fn add_diagnostics(&self, db: &dyn HirDatabase, sink: &mut DiagnosticSink<'_>) {
+        self.diagnostics.iter().for_each(|it| it.add_to(db, sink));
+    }
+
+    /// Returns all implementations defined in this instance.
+    pub fn all_impls(&self) -> impl Iterator<Item = ImplId> + '_ {
+        self.map.values().flatten().copied()
+    }
+
+    // Returns all implementations defined for the specified type.
+    pub fn for_self_ty(&self, self_ty: &Ty) -> &[ImplId] {
+        match self_ty.interned() {
+            TyKind::Struct(s) => self.map.get(&s.id).map_or(&[], AsRef::as_ref),
+            _ => &[],
+        }
+    }
+}
+
+impl InherentImplsDiagnostics {
+    fn add_to(&self, db: &dyn HirDatabase, sink: &mut DiagnosticSink<'_>) {
+        match self {
+            InherentImplsDiagnostics::LowerDiagnostic(impl_id, diag) => {
+                let impl_data = db.impl_data(*impl_id);
+                let file_id = impl_id.lookup(db.upcast()).id.file_id;
+                diag.add_to(db, file_id, &impl_data.type_ref_source_map, sink);
+            }
+            InherentImplsDiagnostics::InvalidSelfTy(impl_id) => sink.push(InvalidSelfTyImpl {
+                impl_: impl_id
+                    .lookup(db.upcast())
+                    .source(db.upcast())
+                    .as_ref()
+                    .map(AstPtr::new),
+            }),
+            InherentImplsDiagnostics::IncoherentType(impl_id) => sink.push(IncoherentImpl {
+                impl_: impl_id
+                    .lookup(db.upcast())
+                    .source(db.upcast())
+                    .as_ref()
+                    .map(AstPtr::new),
+            }),
+            InherentImplsDiagnostics::DuplicateDefinitions(first, second) => {
+                sink.push(DuplicateDefinition {
+                    definition: assoc_item_syntax_node_ptr(db.upcast(), second),
+                    first_definition: assoc_item_syntax_node_ptr(db.upcast(), first),
+                    name: assoc_item_name(db.upcast(), first),
+                });
+            }
+        }
+    }
+
+    fn module_id(&self, db: &dyn DefDatabase) -> ModuleId {
+        match self {
+            InherentImplsDiagnostics::LowerDiagnostic(impl_id, _)
+            | InherentImplsDiagnostics::InvalidSelfTy(impl_id)
+            | InherentImplsDiagnostics::IncoherentType(impl_id) => impl_id.module(db),
+            InherentImplsDiagnostics::DuplicateDefinitions(_first, second) => second.module(db),
+        }
```

Comment:
Why do you use the `second` module at this point and not the first?

Reply:
The error is about a redefinition of a name. So the second definition is to what the error refers.